File: C++/Problems/Problems/problem_3522.cpp
#include "problem_3522.h"

bool problem_3522::test() {
    std::vector<std::string> instructions {"jump", "add", "add", "jump", "add", "jump"};

    std::vector<int> values { 2, 1, 3, 1, -2, -3 };

    auto expected = 1;

    auto result = calculateScore(instructions, values);

    std::cout << result << std::endl;

    return result == expected;
}

long long problem_3522::calculateScore(std::vector<std::string> &instructions, std::vector<int> &values) {
    int i = 0;
    int n = instructions.size();
    std::vector<bool> visited(n);
    long score = 0;
    while (i >= 0 && i < n && !visited[i])
    {
        if (instructions[i] == "add")
        {
            visited[i] = true;
            score += values[i];
            i++;
        }
        else
        {
            visited[i] = true;
            i += values[i];
        }
    }

    return score;
}

File: Python/Problems/Problems/problem3522.py
from .problem import Problem


class Problem3522(Problem):
    def test(self):
        instructions = ["jump", "add", "add", "jump", "add", "jump"]

        values = [2, 1, 3, 1, -2, -3]

        expected = 1

        result = self.calculateScore(instructions, values)

        print(result)

        return result == expected

    def calculateScore(self, instructions, values):
        i = 0
        n = len(instructions)
        visited = [False] * n
        score = 0
        while 0 <= i < n and not visited[i]:
            if (instructions[i] == "add"):
                visited[i] = True
                score += values[i]
                i += 1
            else:
                visited[i] = True
                i += values[i]

        return score